test/integration/stats_integration_test.cc

```diff
@@ -129,5 +132,28 @@ TEST_P(StatsIntegrationTest, WithTagSpecifierWithFixedValue) {
   EXPECT_EQ(live->tags()[0].value_, "xxx");
 }
 
+TEST_P(StatsIntegrationTest, MemoryLargeClusterSize) {
+  const size_t million = 1000 * 1000;
+  const size_t start_mem = Memory::Stats::totalCurrentlyAllocated();
+  if (start_mem == 0) {
+    // Skip this test for platforms where we can't measure memory.
+    return;
+  }
+
+  config_helper_.addConfigModifier([](envoy::config::bootstrap::v2::Bootstrap& bootstrap) {
+    for (int i = 1; i < 1001; i++) {
+      RELEASE_ASSERT(bootstrap.mutable_static_resources()->clusters_size() >= 1, "");
+      auto* c = bootstrap.mutable_static_resources()->add_clusters();
+      c->set_name(fmt::format("cluster_{}", i));
+    }
+  });
+  initialize();
+
+  const size_t end_mem = Memory::Stats::totalCurrentlyAllocated();
```

Comment:
this is great; but can we express the expectation per-cluster, similar to the logic in the script where we took

  (mem(1001_clusters) - mem(1 cluster)) / 1000

also, to get the zero-stat cluster cost, we just need to add a stat-matcher to the config that rejects everything.

Reply:
I just pushed a commit to test with and without stats. There's a bit of duplication that I can factor out into a shared function.

For the 1_cluster memory, should I hardcode that variable or should we run the test once with 1 cluster and then save that to be used in the 1001 cluster test?

Reply:
I wouldn't hardcode it as that's part of what we're trying to measure, and will change as the software evolves.

I think factoring some common bits out would be great. Maybe a helper function like:
  uint64_t sizeConfirmedWithClusters(int num_clusters, bool allow_stats);